### PR TITLE
adjust topn heap operation when string is dictionary encoded, but not uniquely

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
This PR switches `StringTopNColumnAggregatesProcessor`, the string processor used for heap based topN, to use `scanAndAggregateWithCardinalityKnown` for the case where a column is dictionary encoded, but not uniquely, so avoid duplicate decoding of `utf8` bytes from the segment or recomputing expression virtual columns that occur on lookupName in some types of expression selectors. This is potentially worse for things like `IndexedTable`, where dictionaryId might be the row number, so will still have to access the `aggregatesStore` map for every row, but better for things like extractions and expressions on top of "normal" dictionary encoded columns.

The existing topn tests should provide adequate coverage for this change, but will confirm this with the coverage bot once it runs in travis.

This PR has:
- [x] been self-reviewed.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.

### CODE REVIEW

**Comment by clintropolis** [comment]
>The existing topn tests should provide adequate coverage for this change, but will confirm this with the coverage bot once it runs in travis.

[It looks like I was right:](https://app.travis-ci.com/github/apache/druid/jobs/561537298#L2110)

```
Total diff coverage:
 - lines: 100% (24/24)
 - branches: 100% (16/16)
 - functions: 100% (2/2)
```

File: processing/src/main/java/org/apache/druid/query/topn/types/StringTopNColumnAggregatesProcessor.java [hunk 1]
```diff
@@ -172,6 +180,54 @@ private long scanAndAggregateWithCardinalityKnown(
...
+   * an expression virtual column or some lookup value.
+   *
+   */
+  private long scanAndAggregateWithDictionary(
+      TopNQuery query,
+      Cursor cursor,
+      DimensionSelector selector
+  )
+  {
+    Int2ObjectOpenHashMap<Aggregator[]> aggsCache = new Int2ObjectOpenHashMap<>();
```
Comment (imply-cheddar) [inline processing/src/main/java/org/apache/druid/query/topn/types/StringTopNColumnAggregatesProcessor.java:197]: With the test for `hasDictionary` we could perhaps make this

```
Aggregator[][] aggsCache = new Aggregator[selector.getValueCardinality()][];
```

?

**Review by imply-cheddar — APPROVED** [review]

File: processing/src/main/java/org/apache/druid/query/topn/types/StringTopNColumnAggregatesProcessor.java (hunk 1 above)
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/query/topn/types/StringTopNColumnAggregatesProcessor.java:197]: yeah, that should work :+1:

**Comment by gianm** [comment]
I'm trying to understand where we'd be after this patch. What algorithm will be used in each of the following cases?

- string dimension, no dictionary <-- i.e., what we get from an expression that isn't backed by a single dictionary-coded string column
- string dimension, dictionary coded, unique (1-1 mapping from keys -> values) <-- i.e., what we get from a regular column in a segment, or a dictionary-coded string column plus an ExtractionFn that is `ONE_TO_ONE`
- string dimension, dictionary coded, nonunique <-- i.e., what we get from an expression backed by a single dictionary-coded string column, or a dictionary-coded string column plus an ExtractionFn that is `MANY_TO_ONE`

**Comment by clintropolis** [comment]
> * string dimension, no dictionary <-- i.e., what we get from an expression that isn't backed by a single dictionary-coded string column

this one uses scanAndAggregateWithCardinalityUnknown because those columns which are  `1:*` or `*:*` should not report themselves as dictionary encoded/name lookup possible in advance

> * string dimension, dictionary coded, unique (1-1 mapping from keys -> values) <-- i.e., what we get from a regular column in a segment, or a dictionary-coded string column plus an ExtractionFn that is ONE_TO_ONE
> * string dimension, dictionary coded, nonunique <-- i.e., what we get from an expression backed by a single dictionary-coded string column, or a dictionary-coded string column plus an ExtractionFn that is MANY_TO_ONE, or an IndexedTable

these two cases use `scanAndAggregateWithCardinalityKnown`, prior to this patch the latter case used `scanAndAggregateWithCardinalityUnknown`, which is better for `IndexedTable` (since the dictionaryIds never repeat so always has to still perform the value lookupName and hash table lookup), but far worse for expressions and lookups, which never report their dictionaries as unique currently and potentially have many repeated dictionary ids.

Should I do something to clear up the javadocs?

**Comment by gianm** [comment]
Which cases use HeapBasedTopNAlgorithm and which use PooledTopNAlgorithm?

> Should I do something to clear up the javadocs?

Certainly wouldn't hurt to include a comment in the code that selects the algorithm that describes the rationale for the selection.

**Review by gianm — APPROVED** [review]
LGTM

**Comment by gianm** [comment]
> Which cases use HeapBasedTopNAlgorithm and which use PooledTopNAlgorithm?

I realized this code is all in the context of HeapBasedTopNAlgorithm. It looks good to me.

**Pull request merged**